src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs

```diff
@@ -426,18 +426,16 @@ public async Task<ActionResult> RemovePart(string id, string name)
             return NotFound();
         }
 
-        var settings = partDefinition.GetSettings<ContentSettings>();
-
-        if (settings.IsCodeManaged)
-        {
-            await _notifier.ErrorAsync(H["The \"{0}\" part cannot be removed.", name]);
-        }
-        else
```

Comment:
This was fine like this. Throwing an exception from a services if the consumer did something wrong, as well as not doing that mistake from this UI consumer are both useful (and catching any exception here opens up the possibility to swallow and not log more serious errors too).

Reply:
All its duplicating is an `if (settings.IsCodeManaged)` but what happens afterwards is different.

This here is UI code, it communicates with the user. If the user does something they shouldn't, or if something breaks, it should tell the user about that in a way that the user will understand and can act on.

The service is lower-level code that's interacting with other code, not users.

The service code should indicate if its consumers are using it wrong. An exception for this is appropriate.

The UI code here should not use the service wrong, to begin with, and since it very well knows that the user tries to do 
something they shouldn't, it should prevent that from happening in the first place, and communicate the issue to the user.

The controller thus can remain and display an error message to the user specific to `IsCodeManaged`. Catching any exception is not specific at all, and it would show the "part cannot be removed" message even if e.g. the DB query fails. And relying on exceptions for control flow (a user trying something they are not supposed to is really not an exceptional situation) is not something we should do either.

Reply:
> This here is UI code, it communicates with the user. If the user does something they shouldn't, or if something breaks, it should tell the user about that in a way that the user will understand and can act on.
>The service is lower-level code that's interacting with other code, not users.

They user should never encounter this exception from the UI because we don't show delete button for these parts.

If there service throw an exception "failed" for any reason, it's reasonable to show a generic message.

if the RemovePartFromTypeAsync failed for any reason, we should let the user know that something happened. The service class is designed for the UI. so I don't think it is a good idea to duplicate the code. the service is designed to wrote the logic in the controller.

Reply:
Swallowing all exceptions and showing the same message regardless of their type won't help users. It's also an issue that since the exception is swallowed, the usual error handlers won't log it either.

So if you see no point in having this check here, then rather the whole try-catch should be removed too.

Reply:
@Piedone done.

Reply:
Thanks. I'll come back when you two with Sebastien are done, because there's a lot of movement here.